design/design-doc-composition-functions-extra-resources.md

```diff
@@ -0,0 +1,1196 @@
+# Composition Functions - Extra Resources
+
+* Owner: Philippe Scorsolini (@phisco)
+* Reviewers: TBD
+* Status: Draft
+
+## Background
+
+Composition Functions extend Crossplane to support a new way of configuring
+how to reconcile a Composite Resource (XR). Each Function is a gRPC server.
+Crossplane sends state to a Function via a `RunFunctionRequest` RPC. The
+Function is intended to return the desired state via a `RunFunctionResponse`
+RPC.
+
+Currently, a Composition author who needs to access information from a resource
+external to their Composition has the following options:
+- [**BETA**] Create an `ObserveOnly` resource from their Composition and
+  reading the required information from its `status`.
+- [**ALPHA**] Create an `EnvironmentConfig`, either manually or from another
+  Composition, with the required information and reference it from their
+  Composition via `spec.environment.environmentConfigs`.
+- Use MR-level referencing, if supported for the source and target of the
+  information, being implemented by the provider.
+
+Each of these strategies has its drawbacks:
+- The creation of `ObserveOnly` resources could result in a significant number
+  of resources being created.
+- The use of `EnvironmentConfigs` requires the Composition author to create and
+  maintain a separate resource, which can be cumbersome and error-prone.
+  However, it has to be said that there is value in the clean separation of concerns
+  between the `EnvironmentConfig`'s producer and consumer(s).
+- Using MR-level referencing is not always possible, as it requires explicit
+  provider support, and hardly ever crosses a single provider's (family) boundaries.
```

Comment:
In practice, all 3 strategies can be currently combined in the same Crossplane Configuration.

EnvironmentConfig keeps the environment-dependant value -> ObserveOnly MR is using the value from EnvConfig to construct external-name for the import -> MR level selectors are used to reference the ObserveOnly MR from another MRs in the same Composition